ipa-core/src/report/hybrid.rs

```diff
@@ -90,28 +115,100 @@ where
             U16,
         >: ArrayLength,
     {
-        let encrypted_oprf_report = EncryptedOprfReport::<BK, V, TS, B>::from_bytes(data)?;
+        let encrypted_oprf_report =
+            EncryptedOprfReport::<BK, V, TS, Bytes>::try_from(self.bytes.clone())?;
         let oprf_report = encrypted_oprf_report.decrypt(key_registry)?;
         match oprf_report.event_type {
-            EventType::Source => Ok(Self::Impression(HybridImpressionReport {
+            EventType::Source => Ok(HybridReport::Impression(HybridImpressionReport {
                 match_key: oprf_report.match_key,
                 breakdown_key: oprf_report.breakdown_key,
             })),
-            EventType::Trigger => Ok(Self::Conversion(HybridConversionReport {
+            EventType::Trigger => Ok(HybridReport::Conversion(HybridConversionReport {
                 match_key: oprf_report.match_key,
                 value: oprf_report.trigger_value,
             })),
         }
     }
 }
 
+impl TryFrom<Bytes> for EncryptedHybridReport {
+    type Error = InvalidReportError;
+
+    fn try_from(bytes: Bytes) -> Result<Self, InvalidReportError> {
+        Ok(EncryptedHybridReport { bytes })
+    }
+}
+
+pub trait UniqueBytes {
+    fn unique_bytes(&self) -> Vec<u8>;
+}
+
+impl UniqueBytes for EncryptedHybridReport {
+    fn unique_bytes(&self) -> Vec<u8> {
+        let start: usize = EncapsulationSize::USIZE;
+        let end: usize = start + 16;
+        self.bytes[start..end].to_vec()
+    }
+}
+
+#[derive(Debug)]
+pub struct UniqueBytesValidator {
+    hash_set: HashSet<Vec<u8>>,
+    incrementer: usize,
+}
+
+impl UniqueBytesValidator {
+    #[must_use]
+    pub fn new(size: usize) -> Self {
+        UniqueBytesValidator {
+            hash_set: HashSet::<Vec<u8>>::with_capacity(size),
+            incrementer: 0,
+        }
+    }
+
+    fn insert(&mut self, value: Vec<u8>) -> bool {
+        let r = self.hash_set.insert(value);
+        if r {
+            self.incrementer += 1;
```

Comment:
I am not sure if this serves the purpose here. If you want to point to the exact location of the duplicate report, this does not work because it does not take into account previous duplicates. Lets say you have [foo, bar, baz, foo, baz]. This correctly points at second `foo`, but baz will have an index 3 instead of 4. 

If you want to detect only first duplicate and stop, then `incrementer` is always `hash_set.len()` I think. 

I would probably just keep a global counter, or make the caller to be responsible to that. In the latter case, you only need to signal back the index inside `check_duplicates` function and point to the item that failed duplicate check. Then caller can find it in its own array and compute the index and/or just point at the ciphertext

Reply:
ahh yea, good point. what we want is an `check_counter` that increments for every check, regardless of success.